src/EventListener/TagListener.php

```diff
@@ -88,14 +105,15 @@ public function onKernelResponse(FilterResponseEvent $event)
             }
         }
 
-        if ($request->isMethodCacheable()) {
+        if ($this->cacheableRule->matches($request, $response)) {
+            // For safe requests (GET and HEAD), set cache tags on response
             $this->symfonyResponseTagger->addTags($tags);
             if (HttpKernelInterface::MASTER_REQUEST === $event->getRequestType()) {
-                // For safe requests (GET and HEAD), set cache tags on response
                 $this->symfonyResponseTagger->tagSymfonyResponse($response);
             }
-        } elseif (count($tags)) {
-            // For non-safe methods, invalidate the tags
+        } elseif (count($tags)
```

Comment:
This cannot be moved to earlier in the method because `$this->symfonyResponseTagger->tagSymfonyResponse($response);` still has to happen: e.g. for when tags are set from Twig. The `testTwigExtension` test fails if we add `if (count($tags)) { return; }` before this.